### PR TITLE
Disallow empty identifiers

### PR DESCRIPTION
**Pull request by martint**

### CODE REVIEW

File: presto-main/src/test/java/io/prestosql/sql/query/TestSessionFunctions.java [hunk 1]
```diff
@@ -45,11 +45,11 @@ public void testCurrentPath()
         }
 
         Session emptyPathSession = testSessionBuilder()
-                .setPath(new SqlPath(Optional.of("\"\"")))
+                .setPath(new SqlPath(Optional.empty()))
                 .build();
 
         try (QueryAssertions queryAssertions = new QueryAssertions(emptyPathSession)) {
-            queryAssertions.assertQuery("SELECT CURRENT_PATH", "SELECT CAST('" + emptyPathSession.getPath().toString() + "' AS VARCHAR)");
+            queryAssertions.assertQuery("SELECT CURRENT_PATH", "VALUES VARCHAR ''");
```
Comment (findepi) [inline presto-main/src/test/java/io/prestosql/sql/query/TestSessionFunctions.java]: `VALUES ''` should be enough.

however, if path is not set, maybe `CURRENT_PATH` should return `NULL`?

File: presto-parser/src/test/java/io/prestosql/sql/parser/TestSqlParserErrorHandling.java [hunk 2]
```diff
@@ -39,85 +39,85 @@
...
-                        "line 1:23: mismatched input '<EOF>'. Expecting: 'WHERE'"},
+                 "line 1:23: mismatched input '<EOF>'. Expecting: 'WHERE'"},
                 {"SELECT * FROM t t x",
-                        "line 1:19: mismatched input 'x'. Expecting: '(', ',', 'CROSS', 'EXCEPT', 'FULL', 'GROUP', 'HAVING', 'INNER', 'INTERSECT', 'JOIN', 'LEFT', 'LIMIT', 'NATURAL', 'ORDER', " +
-                                "'RIGHT', 'TABLESAMPLE', 'UNION', 'WHERE', <EOF>"},
+                 "line 1:19: mismatched input 'x'. Expecting: '(', ',', 'CROSS', 'EXCEPT', 'FULL', 'GROUP', 'HAVING', 'INNER', 'INTERSECT', 'JOIN', 'LEFT', 'LIMIT', 'NATURAL', 'ORDER', " +
+                         "'RIGHT', 'TABLESAMPLE', 'UNION', 'WHERE', <EOF>"},
                 {"SELECT * FROM t WHERE EXISTS (",
-                        "line 1:31: mismatched input '<EOF>'. Expecting: <query>"}};
+                 "line 1:31: mismatched input '<EOF>'. Expecting: <query>"}};
```
Comment (findepi) [inline presto-parser/src/test/java/io/prestosql/sql/parser/TestSqlParserErrorHandling.java]: end this line with `},` and put `};` on the next line -- you won't need to modify this line in order to add another test case later

File: presto-parser/src/test/java/io/prestosql/sql/parser/TestSqlParserErrorHandling.java [hunk 3]
```diff
@@ -117,7 +117,11 @@
                  "line 1:19: mismatched input 'x'. Expecting: '(', ',', 'CROSS', 'EXCEPT', 'FULL', 'GROUP', 'HAVING', 'INNER', 'INTERSECT', 'JOIN', 'LEFT', 'LIMIT', 'NATURAL', 'ORDER', " +
                          "'RIGHT', 'TABLESAMPLE', 'UNION', 'WHERE', <EOF>"},
                 {"SELECT * FROM t WHERE EXISTS (",
-                 "line 1:31: mismatched input '<EOF>'. Expecting: <query>"}};
+                 "line 1:31: mismatched input '<EOF>'. Expecting: <query>"},
+                {"SELECT \"\" FROM t",
+                 "line 1:8: mismatched input '\"'. Expecting: '*', 'ALL', 'DISTINCT', <expression>, <identifier>"},
+                {"SELECT a FROM \"\"",
+                 "line 1:15: mismatched input '\"'. Expecting: '(', 'LATERAL', 'UNNEST', <identifier>"}};
```
Comment (findepi) [inline presto-parser/src/test/java/io/prestosql/sql/parser/TestSqlParserErrorHandling.java]: Would it be possible to produce something more meaningful?
We sometimes accept something invalid in grammar in order to produce a better exc msg. For example, we could allow `""` as an identifier in grammar and check this when converting to AST?

**Review by findepi — APPROVED** [review]
minors

File: presto-parser/src/test/java/io/prestosql/sql/parser/TestSqlParserErrorHandling.java [hunk 4]
```diff
@@ -39,85 +39,89 @@
...
-                                "'RIGHT', 'TABLESAMPLE', 'UNION', 'WHERE', <EOF>"},
+                 "line 1:19: mismatched input 'x'. Expecting: '(', ',', 'CROSS', 'EXCEPT', 'FULL', 'GROUP', 'HAVING', 'INNER', 'INTERSECT', 'JOIN', 'LEFT', 'LIMIT', 'NATURAL', 'ORDER', " +
+                         "'RIGHT', 'TABLESAMPLE', 'UNION', 'WHERE', <EOF>"},
                 {"SELECT * FROM t WHERE EXISTS (",
-                        "line 1:31: mismatched input '<EOF>'. Expecting: <query>"}};
+                 "line 1:31: mismatched input '<EOF>'. Expecting: <query>"},
+                {"SELECT \"\" FROM t",
+                 "line 1:8: mismatched input '\"'. Expecting: '*', 'ALL', 'DISTINCT', <expression>, <identifier>"},
+                {"SELECT a FROM \"\"",
+                 "line 1:15: mismatched input '\"'. Expecting: '(', 'LATERAL', 'UNNEST', <identifier>"}};
```
Comment (findepi) [inline presto-parser/src/test/java/io/prestosql/sql/parser/TestSqlParserErrorHandling.java]: Add a test case `SELECT a FROM "".t`

File: presto-parser/src/test/java/io/prestosql/sql/parser/TestSqlParserErrorHandling.java [hunk 5]
```diff
@@ -39,85 +39,85 @@
     {
         return new Object[][] {
                 {"",
-                        "line 1:1: mismatched input '<EOF>'. Expecting: 'ALTER', 'ANALYZE', 'CALL', 'COMMIT', 'CREATE', 'DEALLOCATE', 'DELETE', 'DESC', 'DESCRIBE', 'DROP', 'EXECUTE', 'EXPLAIN', 'GRANT', " +
```
Comment (electrum) [inline presto-parser/src/test/java/io/prestosql/sql/parser/TestSqlParserErrorHandling.java]: While I agree that this looks nicer in this circumstance, it does not match the code formatter. To double check, I verified against the formatter settings in the codestyle repo, and the results were the same as the current committed version.

This result makes sense, as the formatter is treating the second item as a continuation indent. Putting the first item on the same line as the brace is a special case. Instead, consider the more general case:
```java
return new Object[][] {
        {
                "abc",
                "xyz",
        },
        {
                "foo",
                "bar",
        },
};
```

I also couldn't find any formatter setting that matches your proposed change.

**Review by electrum — CHANGES_REQUESTED** [review]

File: presto-parser/src/test/java/io/prestosql/sql/parser/TestSqlParserErrorHandling.java (hunk 3 above)
Comment (electrum) [reply presto-parser/src/test/java/io/prestosql/sql/parser/TestSqlParserErrorHandling.java]: I like this idea. We could do this in `AstBuilder`:
```java
if (identifier.isEmpty()) {
    throw new ParsingException("Zero-length delimited identifier not allowed", getLocation(context));
}
```
PostgreSQL error message:
```
# select "";
ERROR:  zero-length delimited identifier at or near """"
LINE 1: select "";
```

File: presto-parser/src/test/java/io/prestosql/sql/parser/TestSqlParserErrorHandling.java [hunk 6]
```diff
@@ -117,7 +117,11 @@
                  "line 1:19: mismatched input 'x'. Expecting: '(', ',', 'CROSS', 'EXCEPT', 'FULL', 'GROUP', 'HAVING', 'INNER', 'INTERSECT', 'JOIN', 'LEFT', 'LIMIT', 'NATURAL', 'ORDER', " +
                          "'RIGHT', 'TABLESAMPLE', 'UNION', 'WHERE', <EOF>"},
                 {"SELECT * FROM t WHERE EXISTS (",
-                 "line 1:31: mismatched input '<EOF>'. Expecting: <query>"}};
```
Comment (electrum) [inline presto-parser/src/test/java/io/prestosql/sql/parser/TestSqlParserErrorHandling.java]: We should add a trailing comma and move the final brace to the next line, so that we don't need to modify unrelated lines when adding new test cases.

File: presto-parser/src/test/java/io/prestosql/sql/parser/TestSqlParserErrorHandling.java (hunk 5 above)
Comment (martint) [reply presto-parser/src/test/java/io/prestosql/sql/parser/TestSqlParserErrorHandling.java]: I just reformatted the code with IntelliJ and that’s what I got. That’s odd.

Comment (martint) [reply presto-parser/src/test/java/io/prestosql/sql/parser/TestSqlParserErrorHandling.java]: I found the setting that causes that. It's "align when multiline" under "array initializer" formatting.

Comment (electrum) [reply presto-parser/src/test/java/io/prestosql/sql/parser/TestSqlParserErrorHandling.java]: That settings seems inconsistent with everything else, as we don't align method parameters, etc. For example, it results in this:
```java
int[] abc = new int[] {1, 2,
                       3, 4};
```
If we're going to start using that style, we need to update the codestyle and reformat the code base to match.

Comment (electrum) [reply presto-parser/src/test/java/io/prestosql/sql/parser/TestSqlParserErrorHandling.java]: If that was set in the codestyle, it would look like this:
```xml
<codeStyleSettings language="JAVA">
  <option name="ALIGN_MULTILINE_ARRAY_INITIALIZER_EXPRESSION" value="true" />
</codeStyleSettings>

Comment (martint) [reply presto-parser/src/test/java/io/prestosql/sql/parser/TestSqlParserErrorHandling.java]: Yeah, I’ll remove that commit. We should consider adopting that style for arrays, though. It’s a little different from method parameters, as arrays contain homogeneous elements and the visual structure can help when there are lots of elements. Clearly we don’t want to force one element per line like we do for method args.

File: presto-main/src/test/java/io/prestosql/sql/query/TestSessionFunctions.java (hunk 1 above)
Comment (martint) [reply presto-main/src/test/java/io/prestosql/sql/query/TestSessionFunctions.java]: No, it's not. `''` has type `VARCHAR(0)`, which make the assertion fail due to type not being an exact match.

**Comment by martint** [comment]
@electrum, @findepi, updated.

File: presto-parser/src/test/java/io/prestosql/sql/parser/TestSqlParserErrorHandling.java [hunk 7]
```diff
@@ -117,7 +117,16 @@
...
-                        "line 1:31: mismatched input '<EOF>'. Expecting: <query>"}};
+                        "line 1:31: mismatched input '<EOF>'. Expecting: <query>"},
+                {"SELECT \"\" FROM t",
+                        "line 1:8: Zero-length delimited identifier not allowed"},
+                {"SELECT a FROM \"\"",
+                        "line 1:15: Zero-length delimited identifier not allowed"},
+                {"SELECT a FROM \"\".t",
+                        "line 1:15: Zero-length delimited identifier not allowed"},
+                {"SELECT a FROM \"\".s.t",
+                        "line 1:15: Zero-length delimited identifier not allowed"}
```
Comment (findepi) [inline presto-parser/src/test/java/io/prestosql/sql/parser/TestSqlParserErrorHandling.java]: nit: end the last test case with `,`

File: presto-parser/src/main/java/io/prestosql/sql/tree/Identifier.java [hunk 8]
```diff
@@ -51,6 +51,7 @@ private Identifier(Optional<NodeLocation> location, String value, boolean delimi
         this.value = value;
         this.delimited = delimited;
 
+        checkArgument(!value.isEmpty(), "value must not be empty");
```
Comment (findepi) [inline presto-parser/src/main/java/io/prestosql/sql/tree/Identifier.java]: `"value is empty"`  would be consistent with `"value is null"` which we use when doing null check.

also use `this.value = requireNonNull(value, "value is null");` above

**Review by findepi — APPROVED** [review]

**Review by electrum — APPROVED** [review]
